src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java

```diff
@@ -1,78 +1,298 @@
+import net.minecraft.util.EnumHand;
+import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.World;
+import net.minecraftforge.common.MinecraftForge;
 import net.minecraftforge.fml.common.eventhandler.Cancelable;
+import net.minecraftforge.fml.relauncher.Side;
+
+import static net.minecraftforge.fml.common.eventhandler.Event.Result.DEFAULT;
+import static net.minecraftforge.fml.common.eventhandler.Event.Result.DENY;
 
 /**
  * PlayerInteractEvent is fired when a player interacts in some way.
- * <br>
- * This event is fired whenever a player interacts in
- * Minecraft#func_147121_ag(),
- * NetHandlerPlayServer#processPlayerBlockPlacement(C08PacketPlayerBlockPlacement),
- * ItemInWorldManager#activateBlockOrUseItem(EntityPlayer, World, ItemStack, int, int, int, int, float, float, float),
- * ItemInWorldManager#onBlockClicked(int, int, int, int). <br>
- * <br>
- * This event is fired via the {@link ForgeEventFactory#onPlayerInteract(EntityPlayer, Action, BlockPos, EnumFacing)}.
- * <br>
- * {@link #action} contains the Action the player performed durin this interaction. <br>
- * {@link #pos} contains the coordinate of where this event occurred.<br>
- * {@link #face} contains the face of the block that was interacted with. May be null if unknown. <br>
- * {@link #world} contains the world in which this event is occurring. <br>
- * <br>
- * This event is {@link Cancelable}.<br>
- * If this event is canceled, the player does not interact.<br>
- * <br>
- * This event does not have a result. {@link HasResult}<br>
- * <br>
- * This event is fired on the {@link MinecraftForge#EVENT_BUS}.
+ * All subclasses are fired on {@link MinecraftForge#EVENT_BUS}.
+ * See the individual documentation on each subevent for more details.
  **/
-@Cancelable
 public class PlayerInteractEvent extends PlayerEvent
 {
-    public static enum Action
+    private final EnumHand hand;
+    private final ItemStack stack;
+    private final BlockPos pos;
+    private final EnumFacing face;
+
+    private PlayerInteractEvent(EntityPlayer player, EnumHand hand, ItemStack stack, BlockPos pos, EnumFacing face)
+    {
+        super(Preconditions.checkNotNull(player, "Null player in PlayerInteractEvent!"));
+        this.hand = Preconditions.checkNotNull(hand, "Null hand in PlayerInteractEvent!");
+        this.stack = stack;
+        this.pos = Preconditions.checkNotNull(pos, "Null position in PlayerInteractEvent!");
+        this.face = face;
+    }
+
+    /**
+     * This event is fired on both sides whenever a player interacts with an entity.
+     * It has the local position (where on the entity) the player interacted at.
+     * Canceling the event clientside will cause processing to continue to EntityInteract (below),
+     * while canceling serverside will simply do no further processing.
+     */
+    @Cancelable
+    public static class EntityInteractSpecific extends PlayerInteractEvent {
```

Comment:
The formatting for the inner classes varies (`{` on same line instead of new line, etc).